search/utils.hpp

```diff
@@ -3,31 +3,120 @@
+                 DFA const & dfa, ToDo && toDo)
+{
+  using Char = typename Trie::Char;
+  using TrieIt = typename Trie::Iterator;
+  using DFAIt = typename DFA::Iterator;
+  using State = pair<TrieIt, DFAIt>;
+
+  std::queue<State> q;
+
+  {
+    auto it = dfa.Begin();
+    if (it.Rejects())
+      return false;
+    q.emplace(trieStartIt, it);
+  }
+
+  bool found = false;
+
+  while (!q.empty())
+  {
+    auto const p = q.front();
+    q.pop();
+
+    auto const & trieIt = p.first;
+    auto const & dfaIt = p.second;
+
+    if (dfaIt.Accepts())
+    {
+      trieIt.ForEachInNode(toDo);
+      found = true;
+    }
+
+    trieIt.ForEachMove([&](Char const & c, TrieIt const & nextTrieIt) {
+      auto nextDfaIt = dfaIt;
+      nextDfaIt.Move(c);
+      if (!nextDfaIt.Rejects())
+        q.emplace(nextTrieIt, nextDfaIt);
+    });
+  }
+
+  return found;
+}
+
 using TLocales = buffer_vector<int8_t, 3>;
 
+size_t GetMaxErrorsForToken(strings::UniString const & token);
+
 template <typename ToDo>
 void ForEachCategoryType(StringSliceBase const & slice, TLocales const & locales,
                          CategoriesHolder const & categories, ToDo && todo)
 {
   for (size_t i = 0; i < slice.Size(); ++i)
   {
     auto const & token = slice.Get(i);
-    for (size_t j = 0; j < locales.size(); ++j)
-      categories.ForEachTypeByName(locales[j], token, bind<void>(todo, i, _1));
+    for (int8_t const locale : locales)
+      categories.ForEachTypeByName(locale, token, std::bind<void>(todo, i, std::placeholders::_1));
 
     // Special case processing of 2 codepoints emoji (e.g. black guy on a bike).
     // Only emoji synonyms can have one codepoint.
     if (token.size() > 1)
     {
       categories.ForEachTypeByName(CategoriesHolder::kEnglishCode, strings::UniString(1, token[0]),
-                                   bind<void>(todo, i, _1));
+                                   std::bind<void>(todo, i, std::placeholders::_1));
     }
   }
 }
+
+// Unlike ForEachCategoryType which extracts types by a token
+// from |slice| by exactly matching it to a token in the name
+// of a category, in the worst case this function has to loop through the tokens
+// in all category synonyms in all |locales| in order to find a token
+// whose edit distance is close enough to the required token from |slice|.
+template <typename ToDo>
+void ForEachCategoryTypeFuzzy(StringSliceBase const & slice, TLocales const & locales,
+                              CategoriesHolder const & categories, ToDo && todo)
+{
+  using Trie = my::MemTrie<strings::UniString, uint32_t>;
+
+  auto const & trie = categories.GetNameToTypesTrie();
+  auto const & trieRootIt = trie.GetRootIterator();
+  std::set<int8_t> localeSet(locales.begin(), locales.end());
+
+  for (size_t i = 0; i < slice.Size(); ++i)
+  {
+    auto const & token = slice.Get(i);
+    auto const & dfa =
+        strings::LevenshteinDFA(token, 1 /* prefixCharsToKeep */, GetMaxErrorsForToken(token));
```

Comment:
At least, leave a TODO here that we build LevenshteinDFA twice for each token (here and in the Geocoder), not quite expensive but still may be optimized.